pkg/provider/azure.go

```diff
@@ -226,6 +226,8 @@ type Config struct {
 	VmssFlexCacheTTLInSeconds int `json:"vmssFlexCacheTTLInSeconds,omitempty" yaml:"vmssFlexCacheTTLInSeconds,omitempty"`
 	// VmssFlexVMCacheTTLInSeconds sets the cache TTL for vmss flex vms
 	VmssFlexVMCacheTTLInSeconds int `json:"vmssFlexVMCacheTTLInSeconds,omitempty" yaml:"vmssFlexVMCacheTTLInSeconds,omitempty"`
+	// VmssFlexUsingDefaultVMProfile if VMSS Flex VM is provisioned using VMSS Flex default VM Profile, it is set to true. Otherwize, it is set to false.
+	VmssFlexUsingDefaultVMProfile *bool `json:"vmssFlexUsingDefaultVMProfile,omitempty" yaml:"vmssFlexUsingDefaultVMProfile,omitempty"`
```

Comment:
I don't think we need a new config here. Since we would need to list VMSS and VMs for caching, a map of  {key: vmss, value: list of (vmname, computer)} could be built while building the cache. Then later the vmss name for a given node could be found by this map.

Reply:
If the node was there before the cache was built, then what you said works. However, if the node is newly added after the generating the cache ( for example, scaling out), then the cache does not include the info of the node. In this case, we need to first find out which VMSS Flex this node belongs to if there are multiple VMSS Flex in the same cluster, then refresh the cache of the corresponding VMSS Flex.

Reply:
When refreshing caches, VMSS list would also be refreshed

Reply:
Then how to handle the case if the VM is added after the cache is refreshed?

Reply:
the cache need to be refreshed again on this case (it is already the case for vmss uniform, as customer may add a new nodepool at any time)

Reply:
Agree, but refreshing the cache is where the problem is: we need to know which VMSS Flex this new node belongs to. 

- if the VM is provisioned by during VMSS default VM profile, then we can use the computerNamePrefix to find out which vmss Flex this new VM belongs to, similar to how the Unifrom works
- If the VM is provisioned by creating individual VM and attaching to an existing VMSS Flex, then there is no default VM profile in the VMSS level. In this case, we need to send a GET request to find out the VMSS ID of this VM, similar to how AvailabilitySet works

The reason we have a new Config here is we need to know this new VM is provisioned by which of the above approaches, as the way the VM is provisioned will decide how should we check the scale set it belongs to

Reply:
Refer https://github.com/kubernetes-sigs/cloud-provider-azure/blob/aa80641f56a5cfbbfb1404ede15c05352fc6d19d/pkg/provider/azure_vmss.go#L689-L721, iterate all VMs in all VMSS would get the nodeIdentity (iterating is ok because all of them are in memory).

Reply:
(ranging VMs is not part of this func yet, but could be added for vmss flex)

Reply:
Per offline discussion,  If it is not in the cache,  CCM will fresh the caches for each VMSS in the K8s cluster and check again.

Reply:
Code modified as discussion